### PR TITLE
HAST-187: Upgrade to .NET 6 ⭐

### LINKED ISSUE #68: Upgrade to .NET 6
Upgrade the solution to .NET 6.

- Update all submodules. Note that Helpful Libraries was refactored a lot: https://github.com/Lombiq/Helpful-Libraries/issues/118
- Enable and update both `FSharpSamplesMatchesApproved` tests.

### PR DESCRIPTION
**Pull request by Piedone**
HAST-187

Fixes #68

### CODE REVIEW

**Comment by sarahelsaig** [comment]
image
https://www.youtube.com/watch?v=eLeh5fEkrF4

File: Hast.Abstractions/Hast.Vitis.Abstractions/Services/VitisHardwareImplementationComposerBuildProvider.cs [hunk 1]
```diff
@@ -577,13 +583,18 @@ private async Task EnsureDeviceReadyAsync(VitisBuildConfiguration buildConfigura
 
         if (promptBeforeBuild)
         {
+            // This is not logging but a prompt in select cases.
+#pragma warning disable S106 // Standard outputs should not be used directly to log anything
+#pragma warning disable S2228 // Console logging should not be used
```
Comment (sarahelsaig) [inline Hast.Abstractions/Hast.Vitis.Abstractions/Services/VitisHardwareImplementationComposerBuildProvider.cs]: Is it necessary to enable both warnings? Seems redundant.

File: Hast.Abstractions/Hast.Vitis.Abstractions/Services/ZynqHardwareImplementationComposerBuildProvider.cs [hunk 2]
```diff
@@ -89,9 +92,11 @@ public sealed class ZynqHardwareImplementationComposerBuildProvider
         await _buildLogger.ExecuteWithLoggingAsync(vivadoExecutable, vivadoArguments, tmpDirectoryPath);
         MajorProgress("Frequency scaling profile created.");
 
+        const string input = "--input";
```
Comment (sarahelsaig) [inline Hast.Abstractions/Hast.Vitis.Abstractions/Services/ZynqHardwareImplementationComposerBuildProvider.cs]: How come you don't need the same for "--force" and "--info" too?

File: Samples/Hast.Samples.SampleAssembly/GenomeMatcher.cs [hunk 3]
```diff
@@ -233,7 +229,7 @@ private string GetResult(SimpleMemory simpleMemory, string inputOne, string inpu
     {
         var maxInputLength = Math.Max(inputOne.Length, inputTwo.Length);
 
-        var result = string.Empty;
+        var result = new StringBuilder();
```
Comment (sarahelsaig) [inline Samples/Hast.Samples.SampleAssembly/GenomeMatcher.cs]: Looks like you should be able to set the [capacity](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.-ctor?view=net-6.0#system-text-stringbuilder-ctor(system-int32)) to `maxInputLength` to eliminate the overhead of growing the internal list.

File: Samples/Hast.Samples.SampleAssembly/GenomeMatcher.cs [hunk 4]
```diff
@@ -242,9 +238,9 @@ private string GetResult(SimpleMemory simpleMemory, string inputOne, string inpu
             var currentCharBytes = BitConverter.GetBytes(currentChar);
             var chars = Encoding.UTF8.GetChars(currentCharBytes);
 
-            result += chars[0];
+            result.Append(chars[0]);
         }
 
-        return result.Replace("\0", string.Empty);
+        return result.ToString().Replace("\0", string.Empty);
```
Comment (sarahelsaig) [inline Samples/Hast.Samples.SampleAssembly/GenomeMatcher.cs]: Why not check if `chars[0] == '\0'` instead of the replace?

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

**Comment by sarahelsaig** [comment]
Don't you also need to update _Hastlayer.SDK.Client.sln_?

File: Hast.Abstractions/Hast.Vitis.Abstractions/Services/ZynqHardwareImplementationComposerBuildProvider.cs (hunk 2 above)
Comment (Piedone) [reply Hast.Abstractions/Hast.Vitis.Abstractions/Services/ZynqHardwareImplementationComposerBuildProvider.cs]: The analyzer's threshold seems to be at 3, so violation for `"--input"` but not for the others. However, thinking about it, this actually makes the code less readable, so I'm suppressing instead.

**Comment by Piedone** [comment]
I did try it, it worked, and I didn't have any second thoughts. However, it's really weird that it worked, actually, the project should be there.

File: Hast.Abstractions/Hast.Vitis.Abstractions/Services/VitisHardwareImplementationComposerBuildProvider.cs (hunk 1 above)
Comment (Piedone) [reply Hast.Abstractions/Hast.Vitis.Abstractions/Services/VitisHardwareImplementationComposerBuildProvider.cs]: Indeed, and [S106 is the rule that's the current one](https://github.com/SonarSource/sonar-dotnet/issues/1812).

**Review by sarahelsaig — APPROVED** [review]

**Pull request merged**